### ISSUE TITLE
snownet: detect NAT status

### ISSUE THREAD

**Issue by thomaseizinger** [CloudFire-LLC/cloudfire#4060]
As a peer starts up, it doesn't yet know anything about its NAT status. As soon as we are given a set of relays with the first connection, we should be able to determine our NAT status:

If we end up with the same server-reflexive candidate when talking to two different relays, we are **not** behind a restrictive NAT. In that case, we should be able to avoid sending our relay candidates to the other party because we should be able to hole-punch.

We need to think about when to reset this state though:

- When we change networks? (i.e. either told by the client-app or when our server-reflexive candidate changes)
- When new connections start failing?

A nice feature we can build on top of this is letting the portal know about our NAT status. Customers should not deploy gateways to restricted NATs if possible so we may want to send them a message if we detect that we are behind a restricted NAT.

**Comment by thomaseizinger** [comment]
With the changes proposed in https://github.com/firezone/firezone/issues/4548, we can actually detect our NAT status early, i.e. before the first connection because we already receive a list of relays in `init`.

**Comment by thomaseizinger** [comment]
With https://github.com/firezone/firezone/pull/4268, sending of the relay candidates is always delayed. It would be better to only delay it if we think that we aren't behind a restrictive NAT.

**Comment by jamilbk** [comment]
> we should be able to avoid sending our relay candidates to the other party because we should be able to hole-punch.

This is a risky assumption to get wrong, no?

AFAIK a busy firewall or one with custom rules (corporate) might have different port mapping behavior when talking to Relays vs Gateway and using different protocols

**Comment by thomaseizinger** [comment]
> > we should be able to avoid sending our relay candidates to the other party because we should be able to hole-punch.
> 
> This is a risky assumption to get wrong, no?

Yes, I think the safer options is what I mentioned in the last comment: Make the delaying of the relay candidates conditional but send them eventually.

**Comment by jamilbk** [comment]
Ah gotcha, so an optimized approach. Now I follow

**Comment by jamilbk** [comment]
Related: #8191

**Comment by thomaseizinger** [comment]
I don't think we need this. We can infer it from whether any of our connections have a `Relay` component.
